### PR TITLE
PMM-7518: Backup restore

### PR DESCRIPTION
**Pull request by fabio-silva**
What this PR does / why we need it: Adds the restore action to backup artifacts;

Which issue(s) this PR fixes: https://jira.percona.com/browse/PMM-7518

# Implementation
- `RestoreBackupModal` to allow user to pick same / compatible service. When compatible service is picked, the user can select from the dropdown (not  yet implemented, missing API);
- The `RadioButtonGroupField` has less height than other inputs, so style had to be overrode there.

# Screenshots
Screenshot 2021-03-18 at 10 15 40
Screenshot 2021-03-18 at 10 15 58

### CODE REVIEW

File: public/app/features/backup/components/BackupInventory/BackupInventory.service.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
+//@ts-nocheck
```
Comment (tiagomotasantos) [inline public/app/features/backup/components/BackupInventory/BackupInventory.service.ts:1]: Why is this needed?

Comment (fabio-silva) [reply public/app/features/backup/components/BackupInventory/BackupInventory.service.ts]: Oops, must have forgotten it, thanks 👍

**Review by tiagomotasantos — APPROVED** [review]

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts [hunk 2]
```diff
@@ -7,13 +7,23 @@ import { SelectableService } from './AddBackupModal.types';
...
+    const supportedServices: Databases[] = [Databases.mysql, Databases.mongodb];
     const services = await InventoryService.getDbServices();
 
-    // TODO remove this constraint when more DB types are supported
-    if (services.mysql) {
-      result = services.mysql.map(
-        ({ id, name }): SelectableValue<SelectableService> => ({ label: name, value: { id, vendor: Databases.mysql } })
-      );
+    let serviceName: Databases;
+    for (serviceName in services) {
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:14]:
```suggestion
    for (const serviceName in services) {
```

File: public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts [hunk 3]
```diff
@@ -32,6 +32,13 @@ export const BackupInventoryService = {
       })
     );
   },
+  async restore(serviceId: string, locationId: string, artifactId: string): Promise<any> {
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts:35]: Please, add the type of the return value

File: public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts [hunk 4]
```diff
@@ -32,6 +32,13 @@ export const BackupInventoryService = {
...
+    return api.post(`${BASE_URL}/Backups/RestoreBackup`, {
+      service_id: serviceId,
+      location_id: locationId,
+      artifact_id: artifactId,
+    });
+  },
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts:41]:
```suggestion
  async restore(serviceId: string, locationId: string, artifactId: string): Promise<any> => 
    api.post(`${BASE_URL}/Backups/RestoreBackup`, {
      service_id: serviceId,
      location_id: locationId,
      artifact_id: artifactId,
    });
```

File: public/app/percona/backup/components/BackupInventory/BackupInventory.tsx [hunk 5]
```diff
@@ -55,14 +57,36 @@ export const BackupInventory: FC = () => {
...
+
+  const handleClose = () => {
+    setSelectedBackup(null);
+    setRestoreModalVisible(false);
+    setBackupModalVisible(false);
+  };
+
+  const handleRestore = async (serviceId: string, locationId: string, artifactId: string) => {
+    try {
+      await BackupInventoryService.restore(serviceId, locationId, artifactId);
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/BackupInventory.tsx:83]: shouldn't we use the cancel token from the other PR?

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import { css } from 'emotion';
+import { GrafanaTheme } from '@grafana/data';
+
+export const getStyles = ({ colors, typography, spacing }: GrafanaTheme) => ({
+  label: css`
+    display: block;
+    text-align: left;
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: isn't text alignment to the left the default?

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    font-weight: ${typography.weight.semibold};
+    line-height: 1.25;
+    margin: ${spacing.formLabelMargin};
+    padding: ${spacing.formLabelPadding};
+    color: ${colors.formLabel};
+  `,
+  select: css`
+    margin-bottom: ${spacing.xl};
+    div[class$='-input-wrapper'] {
+      padding: 7px 8px;
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:18]: let's try to stick to the `spacing` values from the theme (same few lines below)

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { DATABASE_LABELS } from 'app/percona/shared/core';
+import { Backup } from '../BackupInventory.types';
+import { formatDataModel } from '../BackupInventory.utils';
+import { RestoreBackupFormProps, ServiceTypeSelect } from './RestoreBackupModal.types';
+
+export const toFormProps = ({ vendor, serviceId, serviceName, dataModel }: Backup): RestoreBackupFormProps => ({
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts:6]: Let's use one interface for both params and return value

**Review by nicolalamacchia — CHANGES_REQUESTED** [review]

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts (hunk 2 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:14]: By doing that, we're assuming `serviceName` as a string. Since `services` is of `DBServiceList` type (which is a map using `Databases` as key), that can't be done, because we can't use `string` as `Databases`

Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:14]: (this is a way of solving this linter problem that we have on other places where we're forcibly using `any`)

File: public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts (hunk 3 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts:35]: 👍

File: public/app/percona/backup/components/BackupInventory/BackupInventory.tsx (hunk 5 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/BackupInventory.tsx:83]: Yep, indeed, but it's probably better to wait for that to be merged instead of cherry picking it here, what do you think?

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts (hunk 6 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: This style is being used on other places because we don't have any `Select` on core yet, so we use this when we want to mimic the core's style (e.g. `AddNotificationChannelModal`), so I've just copied it from there (see conversation [here](https://perconacorp.slack.com/archives/G017YQ1T155/p1615981974004500)).

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts (hunk 7 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:18]: Same reason as above

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts (hunk 8 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts:6]: Not sure if I follow. You mean having a single interface instead of `Backup` and `RestoreBackupFormProps` ?
They are different, as the form has a service type field ("SAME" vs "COMPATIBLE") that doesn't mean anything on the outside, for example. 
Also, the form takes a `SelectableValue` with `{ label: serviceName, value: serviceId }` for the service field, which allows to extract the service ID right away when submitting it. That field is not relevant outside of the form as well.

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts (hunk 2 above)
Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:14]: Got it, but anyway avoid explicit `for` loops unless strictly needed. In this case it looks to me that it can become `services.map((service: Databases) ...)` or `Object.keys(services).map` etc.

Furthermore always use `const` instead of `let`, unless `let` is really needed.

File: public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts (hunk 4 above)
Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts]: right, sorry, I missed that, let's keep it as it is for now

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts (hunk 6 above)
Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: Can you be more specific? Where did you copy it from? If this code is here to stay, let's make it consistent with our standards anyway. What if we create our Select, I'm pretty sure we won't remember that this `label` style is related to the select component when we'll have to remove it :)

Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: It's both on `AddNotificationChannelModal.styles.ts` and `AddAlertRuleModal.styles.ts` (this one seems was even done by you)

Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: Check here: https://github.com/percona-platform/grafana/pull/26/files#diff-b6533a50f97fa4cf6925f00513bcf0e933c3fec0f922cf28811c47ccdbc49c4dR5

File: public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts (hunk 4 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/BackupInventory.service.ts]: 👍

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts (hunk 8 above)
Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts:6]:
```typescript
type ToFormProps = (props: Backup) => RestoreBackupFormProps;
```

```suggestion
export const toFormProps: ToFormProps = ({ vendor, serviceId, serviceName, dataModel }) => ({
```

something like this

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts (hunk 2 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:14]: copy that

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts (hunk 6 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: Regarding the replacement later on: both the `Select` and the `label` are wrapped within a `Field` component, so replacing it should leave no label behind.
Whan we can do for now is to create a component that just renders that `Field` with the label and a slot for children, which might be `Select` or other thing

Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: My point is that if this code is getting duplicated like that, I'd prefer to have it in one place. That might be outside the scope of this PR though. Still, the fact that some code is already in our codebase doesn't mean it's perfect, we should always review it anyway :)

Again, if it's going to be removed soon, ignore this, especially considering it's a minor thing :)

Comment (nicolalamacchia) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: > Regarding the replacement later on: both the Select and the label are wrapped within a Field component, so replacing it should leave no label behind.

That's true for the React code, but not necessarily for CSS :) The linter wouldn't complain about unused variables in that case, since those are object elements.

Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.styles.ts:7]: I agree that this should be centralised 👍  
Let's take care of that on other issue: https://jira.percona.com/browse/PMM-7922

File: public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts (hunk 8 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/RestoreBackupModal/RestoreBackupModal.utils.ts]: got it 👍

**Review by tiagomotasantos — DISMISSED** [review]

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts [hunk 9]
```diff
@@ -7,14 +7,23 @@ import { SelectableService } from './AddBackupModal.types';
...
 
-    // TODO remove this constraint when more DB types are supported
-    if (services.mysql) {
-      result = services.mysql.map(
-        ({ id, name }): SelectableValue<SelectableService> => ({ label: name, value: { id, vendor: Databases.mysql } })
-      );
-    }
+    Object.keys(services).forEach((serviceName: Databases) => {
+      if (supportedServices.includes(serviceName) && services[serviceName]) {
+        const newServices = services[serviceName] || [];
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:15]: nit: use `??` instead of `||`, if you're just using it as a fallback for `undefined` or `null`

**Review by nicolalamacchia — APPROVED** [review]

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts [hunk 10]
```diff
@@ -7,14 +7,23 @@ import { SelectableService } from './AddBackupModal.types';
...
+    const supportedServices: Databases[] = [Databases.mysql, Databases.mongodb];
     const services = await InventoryService.getDbServices();
 
-    // TODO remove this constraint when more DB types are supported
-    if (services.mysql) {
-      result = services.mysql.map(
-        ({ id, name }): SelectableValue<SelectableService> => ({ label: name, value: { id, vendor: Databases.mysql } })
-      );
-    }
+    Object.keys(services).forEach((serviceName: Databases) => {
```
Comment (nicolalamacchia) [inline public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts:13]: Use a `reduce` here instead of declaring an array variable and looping to push values to it

**Review by nicolalamacchia — CHANGES_REQUESTED** [review]

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts (hunk 9 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts]: 👍

File: public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts (hunk 10 above)
Comment (fabio-silva) [reply public/app/percona/backup/components/BackupInventory/AddBackupModal/AddBackupModal.service.ts]: sure 👍

**Review by nicolalamacchia — APPROVED** [review]
